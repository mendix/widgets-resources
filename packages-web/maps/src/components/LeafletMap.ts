import { Component, createElement } from "react";
import {
    FeatureGroup,
    LatLngLiteral,
    LeafletEvent,
    Map,
    Marker,
    TileLayerOptions,
    icon,
    tileLayer
} from "leaflet";
import * as classNames from "classnames";
import ReactResizeDetector from "react-resize-detector";

import { Container, MapUtils } from "../utils/namespace";
import Utils from "../utils/Utils";
import { Alert } from "./Alert";
type MapProps = Container.MapProps;
type DataSourceLocationProps = Container.DataSourceLocationProps;
type Location = Container.Location;
type SharedProps = MapUtils.SharedProps;

const customUrls = Utils.customUrls;
const mapAttr = Utils.mapAttr;

export interface LeafletMapProps extends SharedProps, MapProps {
    onClickMarker?: (event: LeafletEvent, locationAttr: DataSourceLocationProps) => void;
}

export interface LeafletMapState {
    center: LatLngLiteral;
    alertMessage?: string;
    resized: boolean;
}

export class LeafletMap extends Component<LeafletMapProps, LeafletMapState> {
    private leafletNode?: HTMLDivElement;
    private defaultCenterLocation: LatLngLiteral = { lat: 51.9107963, lng: 4.4789878 };
    private map?: Map;
    private markerGroup = new FeatureGroup();
    private readonly onResizeHandle = this.onResize.bind(this);

    readonly state: LeafletMapState = {
        center: this.defaultCenterLocation,
        resized: false
    };

    render() {
        return createElement("div", {
                className: classNames("widget-maps", this.props.className),
                style: { ...this.props.divStyles , ...Utils.getDimensions(this.props) }
            },
            createElement(Alert, {
                bootstrapStyle: "danger",
                className: "widget-leaflet-maps-alert leaflet-control"
            }, this.state.alertMessage),
            createElement("div",
                {
                    className: "widget-leaflet-maps-wrapper"
                },
                createElement("div", {
                    className: "widget-leaflet-maps",
                    ref: this.getRef
                }),
                createElement(ReactResizeDetector, {
                    handleWidth: true,
                    handleHeight: true,
                    refreshRate: 100,
                    refreshMode: "throttle",
                    onResize: this.onResizeHandle
                })
            )
        );
    }

    componentDidMount() {
        if (this.leafletNode && !this.map) {
            this.map = new Map(this.leafletNode, {
                    scrollWheelZoom: this.props.optionScroll,
                    zoomControl: this.props.optionZoomControl,
                    attributionControl: this.props.attributionControl,
                    zoom: this.props.zoomLevel,
                    minZoom: 2,
                    // Work around for page scroll down to botom on first click on map in Chrome and IE
                    // https://github.com/Leaflet/Leaflet/issues/5392
                    keyboard: false,
                    dragging: this.props.optionDrag
                })
                .addLayer(this.setTileLayer());
            if (this.props.inPreviewMode) {
                this.setDefaultCenter(this.props);
            }
        }
    }

    componentWillReceiveProps(nextProps: LeafletMapProps) {
        if (nextProps.alertMessage !== this.props.alertMessage) {
            this.setState({ alertMessage: nextProps.alertMessage });
        }
        if (this.props.allLocations !== nextProps.allLocations) {
            this.setDefaultCenter(nextProps);
            this.setState({ resized: false });
        }
    }

    componentDidUpdate() {
        if (this.map && !this.props.fetchingData) {
            this.map.panTo(this.state.center);
        }
    }

    componentWillUnmount() {
        if (this.map) {
            this.map.remove();
        }
    }

    private onResize() {
        // When map is placed on the non default tab, the maps has no size.
        // Therefor it need to update on the first view.
        if (this.map && !this.state.resized) {
            this.setDefaultCenter(this.props);
            this.setBounds();
            this.setState({ resized: true });
        }
    }

    private getRef = (node: HTMLDivElement) => {
        this.leafletNode = node;
    }

    private setTileLayer = () => {
        const { mapProvider, mapsToken } = this.props;

        let urlTemplate = "";
        let mapAttribution = "";
        if (mapProvider === "mapBox") {
            urlTemplate = customUrls.mapbox + mapsToken;
            mapAttribution = mapAttr.mapboxAttr;
        } else if (mapProvider === "hereMaps" && mapsToken && mapsToken.indexOf(",") > 0) {
            const splitToken = mapsToken.split(",");
            urlTemplate = customUrls.hereMaps + `app_id=${splitToken[0]}&app_code=${splitToken[1]}`;
            mapAttribution = mapAttr.hereMapsAttr;
        } else {
            urlTemplate = customUrls.openStreetMap;
            mapAttribution = mapAttr.openStreetMapAttr;
        }

        return tileLayer(urlTemplate, {
            attribution: mapAttribution,
            id: mapProvider === "mapBox" ? "mapbox.streets" : undefined
        } as TileLayerOptions);
    }

    private setDefaultCenter = (props: LeafletMapProps) => {
        if (!props.fetchingData) {
            this.renderMarkers(props.allLocations);
        }
    }

    private renderMarkers = (locations?: Location[]) => {
        this.markerGroup.clearLayers();
        if (locations && locations.length) {
            locations.forEach(location => this.createMarker(location).then(marker =>
                this.markerGroup.addLayer(marker.on("click", event => {
                    if (this.props.onClickMarker && location.locationAttr && location.locationAttr.onClickEvent !== "doNothing") {
                        this.props.onClickMarker(event, location.locationAttr);
                    }
                }))).then(layer => this.map ? this.map.addLayer(layer) : undefined)
                .catch(reason => this.setState({ alertMessage: `${reason}` }))
            );
            this.setBounds();
        } else if (this.map) {
            this.map.removeLayer(this.markerGroup);
        }
    }

    private setBounds = () => {
        const { defaultCenterLatitude, defaultCenterLongitude, autoZoom, zoomLevel } = this.props;
        setTimeout(() => {
            if (this.map && this.markerGroup) {
                const bounds = this.markerGroup.getBounds();
                if (bounds.isValid()) {
                    this.map.fitBounds(this.markerGroup.getBounds(), { animate: false }).invalidateSize();
                }
                if (!autoZoom) {
                    if (defaultCenterLatitude && defaultCenterLongitude) {
                        this.map.panTo({
                            lat: parseFloat(defaultCenterLatitude),
                            lng: parseFloat(defaultCenterLongitude)
                        }, { animate: false });
                    }
                    this.map.setZoom(zoomLevel);
                }
            }
        }, 0);
    }

    private createMarker = (location: Location): Promise<Marker> =>
        new Promise((resolve) => {
            const { latitude, longitude, url } = location;
            if (url) {
                resolve(
                    new Marker([ Number(latitude), Number(longitude) ]).setIcon(icon({
                        iconUrl: url,
                        iconSize: [ 38, 95 ],
                        iconAnchor: [ 22, 94 ],
                        className: "marker"
                    }))
                );
            } else {
                resolve(new Marker([ Number(latitude), Number(longitude) ]));
            }
        })
}
