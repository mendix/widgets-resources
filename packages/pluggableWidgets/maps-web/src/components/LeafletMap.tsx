import { createElement, ReactElement, useEffect, useRef } from "react";
import { Map, Marker as MarkerComponent, Popup, TileLayer } from "react-leaflet";
import classNames from "classnames";
import { SharedProps } from "../../typings/shared";
import { MapProviderEnum } from "../../typings/MapsProps";
import { getDimensions } from "../utils/dimension";
import { translateZoom } from "../utils/zoom";
import { latLngBounds } from "leaflet";
import { baseMapLayer } from "../utils/leaflet";

export interface LeafletProps extends SharedProps {
    mapProvider: MapProviderEnum;
    attributionControl: boolean;
}

export function LeafletMap(props: LeafletProps): ReactElement {
    const map = useRef<Map>();
    const center = { lat: 51.906688, lng: 4.48837 };
    const {
        autoZoom,
        attributionControl,
        className,
        currentLocation,
        locations,
        mapProvider,
        mapsToken,
        optionScroll: scrollWheelZoom,
        optionZoomControl: zoomControl,
        style,
        zoomLevel: zoom,
        optionDrag: dragging
    } = props;

    useEffect(() => {
        if (map.current) {
            const { leafletElement: mapRef } = map.current;
            const bounds = latLngBounds(
                locations
                    .concat(currentLocation ? [currentLocation] : [])
                    .filter(m => !!m)
                    .map(m => [m.latitude, m.longitude])
            );
            if (bounds.isValid()) {
                if (autoZoom) {
                    mapRef.fitBounds(bounds, { padding: [0.5, 0.5] }).invalidateSize();
                } else {
                    mapRef.panTo(bounds.getCenter(), { animate: false });
                }
            }
        }
    }, [map.current, locations, currentLocation, autoZoom]);

    return (
        <div className={classNames("widget-maps", className)} style={{ ...style, ...getDimensions(props) }}>
            <div className="widget-leaflet-maps-wrapper">
                <Map
                    attributionControl={attributionControl}
                    center={center}
                    className="widget-leaflet-maps"
                    dragging={dragging}
                    maxZoom={18}
                    minZoom={1}
                    ref={ref => {
                        if (ref && ref !== map.current) {
                            map.current = ref;
                        }
                    }}
                    scrollWheelZoom={scrollWheelZoom}
                    zoom={autoZoom ? translateZoom("city") : zoom}
                    zoomControl={zoomControl}
                >
                    <TileLayer
                        {...baseMapLayer(mapProvider, mapsToken)}
                        id={mapProvider === "mapBox" ? "mapbox.streets" : undefined}
                    />
                    {locations
                        .concat(currentLocation ? [currentLocation] : [])
                        .filter(m => !!m)
                        .map((marker, index) => (
                            <MarkerComponent
                                key={`marker_${index}`}
                                position={{ lat: marker.latitude, lng: marker.longitude }}
                                onclick={marker.title ? undefined : marker.onClick}
                                interactive={!!marker.title || !!marker.onClick}
                                title={marker.title}
                            >
                                {marker.title && (
                                    <Popup>
                                        <span
                                            style={{ cursor: marker.onClick ? "pointer" : "none" }}
                                            onClick={marker.onClick}
                                        >
                                            {marker.title}
                                        </span>
                                    </Popup>
                                )}
                            </MarkerComponent>
                        ))}
                </Map>
            </div>
        </div>
    );
}
