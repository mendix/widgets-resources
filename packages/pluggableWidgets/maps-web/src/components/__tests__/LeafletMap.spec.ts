import { shallow, ShallowWrapper } from "enzyme";
import { createElement } from "react";
import { LeafletMap, LeafletProps } from "../LeafletMap";

describe("Leaflet maps", () => {
    const defaultProps: LeafletProps = {
        attributionControl: false,
        autoZoom: true,
        className: "",
        currentLocation: undefined,
        height: 75,
        heightUnit: "pixels",
        locations: [],
        mapProvider: "openStreet",
        mapsToken: "",
        optionDrag: true,
        optionScroll: true,
        optionZoomControl: true,
        showCurrentLocation: false,
        style: {},
        width: 50,
        widthUnit: "percentage",
        zoomLevel: 10
    };

    const renderLeafletMap = (props: LeafletProps): ShallowWrapper<LeafletProps, any> =>
        shallow(createElement(LeafletMap, props));

    it("renders a map with right structure", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "pixels"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with pixels renders structure correctly", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            heightUnit: "pixels",
            widthUnit: "pixels"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of width and height units renders the structure correctly", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            heightUnit: "percentageOfWidth",
            widthUnit: "percentage"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with percentage of parent units renders the structure correctly", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            heightUnit: "percentageOfParent",
            widthUnit: "percentage"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with HERE maps as provider", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            mapProvider: "hereMaps"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with MapBox maps as provider", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            mapProvider: "mapBox"
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with attribution", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            attributionControl: true
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with markers", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            locations: [
                {
                    title: "Mendix HQ",
                    latitude: 51.906688,
                    longitude: 4.48837,
                    url: "image:url"
                },
                {
                    title: "Gementee Rotterdam",
                    latitude: 51.922823,
                    longitude: 4.479632,
                    url: "image:url"
                }
            ]
        });

        expect(leafletMaps).toMatchSnapshot();
    });

    it("renders a map with current location", () => {
        const leafletMaps = renderLeafletMap(defaultProps);
        leafletMaps.setProps({
            showCurrentLocation: true,
            currentLocation: {
                latitude: 51.906688,
                longitude: 4.48837,
                url: "image:url"
            }
        });

        expect(leafletMaps).toMatchSnapshot();
    });
});
