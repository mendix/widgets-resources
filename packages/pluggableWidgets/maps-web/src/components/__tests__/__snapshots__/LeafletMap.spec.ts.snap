// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaflet maps renders a map with HERE maps as provider 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="Map &copy; 1987-2020 <a href='https://developer.here.com'>HERE</a>"
        url="https://2.base.maps.cit.api.here.com/maptile/2.1/maptile/newest/normal.day/{z}/{x}/{y}/256/png8"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with MapBox maps as provider 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="Map data &copy; <a href='https://www.openstreetmap.org/'>OpenStreetMap</a> contributors, <a href='https://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>, Imagery Â© <a href='https://www.mapbox.com/'>Mapbox</a>"
        id="mapbox.streets"
        url="https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with attribution 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={true}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with current location 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
      <ForwardRef(Leaflet(Marker))
        interactive={false}
        key="marker_0"
        position={
          Object {
            "lat": 51.906688,
            "lng": 4.48837,
          }
        }
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with markers 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
      <ForwardRef(Leaflet(Marker))
        interactive={true}
        key="marker_0"
        position={
          Object {
            "lat": 51.906688,
            "lng": 4.48837,
          }
        }
        title="Mendix HQ"
      >
        <ForwardRef(Leaflet(Popup))>
          <span
            style={
              Object {
                "cursor": "none",
              }
            }
          >
            Mendix HQ
          </span>
        </ForwardRef(Leaflet(Popup))>
      </ForwardRef(Leaflet(Marker))>
      <ForwardRef(Leaflet(Marker))
        interactive={true}
        key="marker_1"
        position={
          Object {
            "lat": 51.922823,
            "lng": 4.479632,
          }
        }
        title="Gementee Rotterdam"
      >
        <ForwardRef(Leaflet(Popup))>
          <span
            style={
              Object {
                "cursor": "none",
              }
            }
          >
            Gementee Rotterdam
          </span>
        </ForwardRef(Leaflet(Popup))>
      </ForwardRef(Leaflet(Marker))>
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with percentage of parent units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75%",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with percentage of width and height units renders the structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "auto",
      "paddingBottom": "37.5%",
      "width": "50%",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with pixels renders structure correctly 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "75px",
      "width": "50px",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;

exports[`Leaflet maps renders a map with right structure 1`] = `
<div
  className="widget-maps"
  style={
    Object {
      "height": "37.5px",
      "width": "50px",
    }
  }
>
  <div
    className="widget-leaflet-maps-wrapper"
  >
    <Map
      attributionControl={false}
      center={
        Object {
          "lat": 51.906688,
          "lng": 4.48837,
        }
      }
      className="widget-leaflet-maps"
      dragging={true}
      maxZoom={18}
      minZoom={1}
      scrollWheelZoom={true}
      zoom={10}
      zoomControl={true}
    >
      <ForwardRef(Leaflet(TileLayer))
        attribution="&copy; <a href='https://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.osm.org/{z}/{x}/{y}.png"
      />
    </Map>
  </div>
</div>
`;
